### ISSUE TITLE
Location Restoration Idea from TrackableHit

### ISSUE THREAD

**Issue by ReiiYuki** [google-ar/arcore-unity-sdk#187]
Hi

I am currently working on my project which use ARCore in our main application. 

I have an idea to use TrackableHit from the real world as the reference point.

Then I create method to convert position from ARCore to position in my reference point coordinate system which I call it **"Real World Position"**.

Then I store that position in server and trying to restore it back by converting **"Real World Position"** to position in ARCore. 

I aim that after the restoration the object have to be place in the same position.

However I close my application and open it again, the position is seem a bit error in x z but for y is seem significant error. I'm not sure my idea is correct or not.

The other thing that I still struck on is rotation restoration. 

If you have any suggestion or idea please comment.

Thanks!

The following is my code:

```C#
namespace ARTag
{
    using UnityEngine;

    public class CoordinateUtils
    {

        public static Vector3 TransformPosition(Vector3 obj, Vector3 pos, Vector3 rot, bool isInverse = false)
        {
            Vector3 ori = obj - pos;

            float x = ori.x;
            float y = ori.y;
            float z = ori.z;

            float a = Mathf.Deg2Rad * rot.x;
            float b = Mathf.Deg2Rad * rot.y;
            float g = Mathf.Deg2Rad * rot.z;

            if (isInverse)
            {
                a = -a;
                b = -b;
                g = -g;
            }

            float[,] rx = new float[,] {
            { 1, 0, 0 },
            { 0, Mathf.Cos(a), -Mathf.Sin(a) },
            { 0, Mathf.Sin(a), Mathf.Cos(a) }
        };
            float[,] ry = new float[,] {
            { Mathf.Cos(b), 0, Mathf.Sin(b) },
            { 0, 1, 0 },
            { -Mathf.Sin(b), 0, Mathf.Cos(b) }
        };
            float[,] rz = new float[,] {
            { Mathf.Cos(g), -Mathf.Sin(g), 0 },
            { Mathf.Sin(g), Mathf.Cos(g), 0 },
            { 0, 0, 1 }
        };

            if (!isInverse)
            {
                return MatrixMult(rx, MatrixMult(ry, MatrixMult(rz, ori)));
            }
            else
            {
                return MatrixMult(rz, MatrixMult(ry, MatrixMult(rx, ori)));
            }
        }

        public static Vector3 TransformRevRotation(Quaternion rot, Quaternion refRot)
        {
            return rot.eulerAngles - refRot.eulerAngles;
        }

        public static Vector3 MatrixMult(float[,] rotationMatrix, Vector3 position)
        {
            float[] positionMatrix = new float[3];
            positionMatrix[0] = position.x;
            positionMatrix[1] = position.y;
            positionMatrix[2] = position.z;

            float[] output = new float[] { 0, 0, 0 };
            for (int row = 0; row < 3; row++)
            {
                output[row] = 0;
                for (int col = 0; col < 3; col++)
                {
                    output[row] += rotationMatrix[row, col] * positionMatrix[col];
                }
            }

            return new Vector3(output[0], output[1], output[2]);
        }
    }

}
```

```C#
public Vector3 GetRealWorldPosition(Vector3 position)
        {
            return CoordinateUtils.TransformPosition(position, refPointPosition, refPointRotation.eulerAngles, true);
        }

        public Vector3 GetVirtualPosition(Vector3 position)
        {
            Vector3 zero = CoordinateUtils.TransformPosition(Vector3.zero, refPointPosition, refPointRotation.eulerAngles, true);
            return CoordinateUtils.TransformPosition(position, zero, refPointRotation.eulerAngles);
        }

        public Quaternion GetRealWorldRotation(Quaternion rotation)
        {
            Vector3 eulerAngle = CoordinateUtils.TransformRevRotation(rotation, refPointRotation);
            return Quaternion.Euler(eulerAngle.x, eulerAngle.y, eulerAngle.z);
        }

        public Quaternion GetVirtualRotation(Quaternion rotation)
        {
            Vector3 eulerAngle = CoordinateUtils.TransformRevRotation(refPointRotation, rotation);
            return Quaternion.Euler(eulerAngle.x, eulerAngle.y, eulerAngle.z);
        }
```

ps. The idea in code is doing translation of axes, then do the rotation of axes by rotation matrix.

ref: 
- https://en.wikipedia.org/wiki/Translation_of_axes
- https://en.wikipedia.org/wiki/Rotation_of_axes
- https://en.wikipedia.org/wiki/Rotation_matrix

**Comment by iBicha** [comment]
Cloud Anchors should help you now.

**Comment by pablisho** [comment]
Hi, **Cloud Anchors** has been released in [ARCore SDK for Unity v1.2.0](https://github.com/google-ar/arcore-unity-sdk/releases/tag/v1.2.0). Cloud Anchors would live up to 24 hs.
Does that help with your local restoration?

**Comment by pablisho** [comment]
Hi, I'm closing this due to inactivity. Please feel free to reopen if you have questions or feature requests about this that cannot be accomplished with Cloud Anchors.
